napari/plugins/_npe2.py

```diff
@@ -125,19 +144,132 @@ def get_widget_contribution(
     return None
 
 
-def populate_qmenu(menu: QMenu, menu_key: str):
-    """Populate `menu` from a `menu_key` offering in the manifest."""
-    # TODO: declare somewhere what menu_keys are valid.
+def _build_menus(entries: Iterable[MenuEntry]) -> List[MenuItem]:
+    """Build napari native menus from npe2 menu entries.
+
+    Parameters
+    ----------
+    entries : iterable of npe2.manifest.contributions.MenuItem
+        npe2 menu entries to build from.
+
+    Returns
+    -------
+    menu_item : list of napari.components.menu.MenuItem
+        Built napari native menu items.
+
+    Raises
+    ------
+    TypeError
+        If any menu entry is not a menu command or submenu.
+    """
+    return [build_menu_item(entry) for entry in entries]
+
+
+@functools.cache
+def build_submenu(submenu_key: str) -> Menu:
+    """Build a napari native menu from an npe2 submenu entry given its key.
+
+    Parameters
+    ----------
+    submenu_key : str
+        Submenu key to query the npe2 plugin manger for.
+
+    Returns
+    -------
+    menu_item : napari.components.menu.Menu
+        Built napari native menu.
+
+    Raises
+    ------
+    TypeError
+        If any submenu child is not a menu command or submenu.
+    """
+    pm = npe2.PluginManager.instance()
+
+    entry = pm.get_submenu(submenu_key)
+    children = _build_menus(entry.contents)
+    return Menu(id=entry.id, label=entry.label, children=children)
+
+
+def build_menu_item(menu_entry: MenuEntry) -> MenuItem:
+    """Build a napari native menu item from an npe2 menu entry.
+
+    Parameters
+    ----------
+    menu_entry : npe2.manifest.contributions.MenuItem
+        npe2 menu entry to build from.
+
+    Returns
+    -------
+    menu_item : napari.components.menu.MenuItem
+        Built napari native menu item.
+
+    Raises
+    ------
+    TypeError
+        If the menu entry is not a menu command or submenu.
+    """
     pm = npe2.PluginManager.instance()
-    for item in pm.iter_menu(menu_key):
-        if isinstance(item, Submenu):
-            subm_contrib = pm.get_submenu(item.submenu)
-            subm = menu.addMenu(subm_contrib.label)
-            populate_qmenu(subm, subm_contrib.id)
-        else:
-            cmd = pm.get_command(item.command)
-            action = menu.addAction(cmd.title)
-            action.triggered.connect(lambda *args: cmd.exec(args=args))  # type: ignore
+
+    if isinstance(menu_entry, MenuCommand):
+        command = pm.get_command(menu_entry.command)
+        return ActionMenuItem(
+            id=command.id, label=command.title, action=command.exec
```

Comment:
we _might_ want to wrap that `command.exec` inside of `napari.utils._injection.inject_napari_dependencies`?  This is still a somewhat new feature.  But if `command.exec` takes parameters, then calling the action will fail.  This is a generally challenging problem, but `inject_napari_dependencies` will handle the most common basic napari type annotations (i.e. a function that takes just a single `viewer`, or `layer` parameter)

Reply:
it died :(

```
napari/plugins/_npe2.py:220: in build_menu_item
    action=inject_napari_dependencies(command.exec),
napari/utils/_injection.py:102: in inject_napari_dependencies
    hints = napari_type_hints(func)
napari/utils/_injection.py:56: in napari_type_hints
    return get_type_hints(
../../.conda/envs/napari/lib/python3.10/site-packages/typing_extensions.py:1298: in get_type_hints
    hint = typing.get_type_hints(
../../.conda/envs/napari/lib/python3.10/typing.py:1870: in get_type_hints
    value = _eval_type(value, globalns, localns)
../../.conda/envs/napari/lib/python3.10/typing.py:329: in _eval_type
    ev_args = tuple(_eval_type(a, globalns, localns, recursive_guard) for a in t.__args__)
../../.conda/envs/napari/lib/python3.10/typing.py:329: in <genexpr>
    ev_args = tuple(_eval_type(a, globalns, localns, recursive_guard) for a in t.__args__)
../../.conda/envs/napari/lib/python3.10/typing.py:327: in _eval_type
    return t._evaluate(globalns, localns, recursive_guard)
../../.conda/envs/napari/lib/python3.10/typing.py:694: in _evaluate
    eval(self.__forward_code__, globalns, localns),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   NameError: name 'CommandRegistry' is not defined

<string>:1: NameError
```

not sure if this is an issue with the mock aspect of the tests or something else